Adding support for Unknown Extra block and appended data after Terminal block

gdesmar opened the pull request.

Matmaus commented on LnkParse3/extra/terminal.py:
```diff
@@ -0,0 +1,33 @@
+    def as_dict(self):
+        tmp = super().as_dict()
+        # Overwrite the usual size with the real appended data length
+        tmp["size"] = len(self.appended_data())
```
Maybe it would be better to override size method with the actual size. Then we do not even need to overwite it here I think.

Matmaus commented on LnkParse3/extra_data.py:
```diff
@@ -36,10 +37,15 @@ def _iter(self):
                 yield cls(indata=data, cp=self.cp)
 
+        # If there is data following the Terminal Block, we should take note of it and tell the user.
+        if len(rest) > 4 and rest[:4] < b"\x00\x00\x00\x04":
```
> In terminal.py, my documentation says BlockSignature == 0x00000000 - 0x00000003, and I was wondering, based on endianness, if it was wrong and should be BlockSignature == 0x00000000 - 0x03000000? Same question would apply, but with more side effects in ExtraData._iter() with if len(rest) > 4 and rest[:4] < b"\x00\x00\x00\x04":, if that check is wrong based on the endianness. I modified it so that it would cover more than only 0x00, but it may cover WAY more now.

Your documentation is correct. About attached line, to be sure, we can do the same what we do with signature - parse it as an integer.
```python
unpack("<I", rest[:4])[0] < 0x00000004
```

Matmaus reviewed (commented):
Thank you for driving this improvement. Most of my remarks are responses to your questions asked in the related issue. I did not realize all the difficulties so it is good you have pointed them out. Also, could you please add two files mentioned in the issue as new tests?

gdesmar commented:
I modified the Unknown block to not use a property, since none of the other block does.
I regenerated the tests to pass, and realized you already had a file with data following the Terminal block, so I added it also as a txt output test.

Regarding text output, I modified the print_lnk_file function to fit the current coding style, but I'm not a fan of hard-coding keys for which there is different behaviour. I went down a very deep rabbit hole to make it all agnostic. It would make those hard-coded list items re-use their parent instead: `Serialized property values` having many `Property` under it, in the agnostic first version, those got transformed from `Property:` to `Serialized property values - <N>:` with <N> being the index, and `Storage:` under `Property store:` to `Property store - <N>:`, thinking that now any list would be handled. I then decided that the extra layers for lists were long and probably not needed and did something similar to yaml where list items have a `-` in front. It's nice looking but the code got a little complicated, I would rather open a new PR if I was to offer it. :D
There is one little tweak I did in the text output, for the nice_id, it will uppercase the key name if it's a hash algorithm, so that SHA256 does not show up as Sha256.

gdesmar commented:
If you were curious to see what difference the agnostic+yaml changes makes in the output (and the code 😨), I uploaded a [new branch](https://github.com/gdesmar/LnkParse3/compare/terminal_block...gdesmar:LnkParse3:agnostic_list_print) with it.
I tried to stick as closely to the original for the early print levels, such that level 1 are **not nice'd** but _has a colon_, level 2 are **not nice'd** and **no colon**, while other levels are _nice'd_ and _has a colon_. It was an interesting journey. 😃

Matmaus commented on LnkParse3/extra/terminal.py:
```diff
@@ -0,0 +1,37 @@
+
+    def as_dict(self):
+        tmp = super().as_dict()
+        tmp["sha256"] = hashlib.sha256(self.appended_data()).hexdigest()
```
Could you please rename `sha256` to `appended_data_sha256`. Similar for `Unknown` block. It will be more clear what has been hashed.

Matmaus commented on tests/test_samples.py:
```diff
@@ -108,8 +108,24 @@ def test_print_shortcut_target_readable(self):
+        self.assertEqual(our, their)
+
+    def test_print_unknown_target(self):
+        with open('tests/samples/unknown_target', 'rb') as indata:
```
Could you please add tests for [sample](https://www.virustotal.com/gui/file/891b561432a26bd71bb67dce82d52bc99e3b45c2d2bf6b714d496ef62b40f2cb) with `Unknown` block?

gdesmar replied:
I'm sorry as it may seem very picky, but I am not supposed to reupload VT samples as they are not free access. 🙁
I have found [this](https://bazaar.abuse.ch/sample/5a3f1d14b9cc4890db64fbc41818d7039f25b0120574dcdec4e20d13e6b2740c) and [this](https://bazaar.abuse.ch/sample/050c65d45e5f21018aa940f0188c4aa1318ac3df865d901f8643ed7ce4a4b52c) sample on Malware Bazaar, which is free for access, but both are 4MB instead of VT's original one at 800KB. I don't know if you have a tool that could be used to forge a sample that would be smaller for the tests? On top of it, I think it's a good idea to have as little real malicious samples stored in a non-defanged state on a public access platform. Let me know what you think! 🙂

Matmaus replied:
Well, I understand your point. Luckily, we can easily corrupt/modify any of test samples. I'll do it once this PR is merged.

Here is output from corrupted `invalid_date2`:
```
EXTRA BLOCKS:
    SPECIAL_FOLDER_LOCATION_BLOCK
        Size: 16
        Special folder id: 37
        Offset: 213
    UNKNOWN_BLOCK
        Block:
            Size: 28
            Extra data sha256: 4f022b4bc7668d870e158010c9877224df1b9b07bd24ef32b731963232b15eb2
            Size: 153
            Extra data sha256: 40ebe6ee6ad3303ae2db241742e857aee523b21553a51e2f6c2a2461f1010879

```
I'll fix print of "Block"  afterwards.

Matmaus reviewed (approved):
LGTM, just two last remarks (also, please fix linting issues so that all the tests will pass).

Regarding you agnostic print approach, I like it. Let's discuss it in the follow up PR.
